pytorch_lightning/callbacks/timer.py

```diff
@@ -0,0 +1,117 @@
+        interval: Determines if the interruption happens on epoch level or mid-epoch.
+            Can be either `epoch` or `step`.
+        verbose: Set this to ``False`` to suppress logging messages.
+
+    Raises:
+        MisconfigurationException:
+            If ``interval`` is not one of the supported choices.
+
+    Example::
+        from pytorch_lightning import Trainer
+        from pytorch_lightning.callbacks import Timer
+
+        # stop training after 12 hours
+        timer = Timer(duration="12:00:00")
+
+        # or provide a datetime.timedelta
+        from datetime import timedelta
+        timer = Timer(duration=timedelta(weeks=1))
+
+        # force training to stop after given time limit
+        trainer = Trainer(callbacks=[timer])
+    """
+
+    INTERVAL_CHOICES = ("epoch", "step")
+
+    def __init__(self, duration: Union[str, timedelta], interval: str = "step", verbose: bool = True):
+        super().__init__()
+        if isinstance(duration, str):
+            hms = datetime.strptime(duration.strip(), "%H:%M:%S")
+            duration = timedelta(hours=hms.hour, minutes=hms.minute, seconds=hms.second)
+        if interval not in self.INTERVAL_CHOICES:
+            raise MisconfigurationException(
+                f"Unsupported parameter value `Timer(interval={interval})`. Possible choices are:"
+                f" {', '.join(self.INTERVAL_CHOICES)}"
+            )
+        self._duration = duration
+        self._interval = interval
+        self._verbose = verbose
+        self._start_time = None
+        self._offset = timedelta()
+
+    @property
+    def start_time(self):
+        return self._start_time
+
+    @property
+    def time_elapsed(self) -> timedelta:
+        if self._start_time is None:
+            return self._offset
+        return datetime.now() - self._start_time + self._offset
+
+    @property
+    def time_remaining(self) -> timedelta:
+        return self._duration - self.time_elapsed
+
+    def on_train_start(self, trainer, *args, **kwargs) -> None:
+        self._start_time = datetime.now()
+
+    def on_train_batch_end(self, trainer, *args, **kwargs) -> None:
+        if self._interval != "step":
+            return
+        self._check_time_remaining(trainer)
+
+    def on_train_epoch_end(self, trainer, *args, **kwargs) -> None:
+        if self._interval != "epoch":
+            return
+        self._check_time_remaining(trainer)
+
+    def on_save_checkpoint(self, trainer, pl_module, checkpoint: Dict[str, Any]) -> Dict[str, Any]:
+        return {
+            "time_elapsed": self.time_elapsed,
+        }
+
+    def on_load_checkpoint(self, callback_state: Dict[str, Any]):
+        self._offset = callback_state.get("time_elapsed", timedelta())
+
+    def _check_time_remaining(self, trainer) -> None:
+        should_stop = self.time_elapsed >= self._duration
+        should_stop = trainer.training_type_plugin.reduce_boolean_decision(should_stop)
```

Comment:
Is it needed ? Can't only rank 0 track the time ?

Reply:
we have to track it on all ranks the same to avoid diverging paths cc @ananthsub 
However, if you want we can broadcast the result instead of reducing.
Then only rank 0 does the decision.

Reply:
all reduce is more conservative for stopping. rank 0 hitting the limit and then broadcasting could let this trigger sooner in case there's some time skew in the ranks.

broadcast is faster though, so let's go with that. we should document this difference here

Reply:
using broadcast now.